### PR TITLE
Added basic screenshot.

### PR DESCRIPTION
**Pull request by sebhz**
Right now the screenshots are placed in the ./screenshot
   directory by default.
   If this directory does not exist, no screenshot will be taken.
   Screenshots are named dumpX.png where X is the number of dump
   taken so far during the session (starting at 0). It would be
   best to create a unique name per session using the current date
   for example.

### CODE REVIEW

File: src/cap32.cpp [hunk 1]
```diff
@@ -3330,7 +3335,25 @@ void set_osd_message(const std::string& message) {
...
+#ifdef HAVE_PNG
+void dumpScreen(void) {
+   static int dump_num=0;
+   struct stat _stat;
 
+   if (stat(CPC.sdump_dir.c_str(), &_stat) == 0) {
+      if (S_ISDIR(_stat.st_mode)) {
+         SDL_Surface* shot;
+         shot = SDL_PNGFormatAlpha(back_surface);
+         std::string dumpPath = CPC.sdump_dir + "/" + std::string("dump") + std::to_string(dump_num) + ".png";
```
Comment (ColinPitrat) [inline src/cap32.cpp]: You could avoid overwriting an existing snapshot (from previous session) and handle incrementing dump_num at the same time here by doing something like:

std::string dumpPath;
do
{
  dumpPath = ...;
  dump_num++;
} while (access(dumpPath.c_str(), F_OK) == 0)

File: src/cap32.cpp [hunk 2]
```diff
@@ -3330,7 +3335,25 @@ void set_osd_message(const std::string& message) {
...
+         LOG_DEBUG("Dumping screen to " + dumpPath);
+         SDL_SavePNG(shot, dumpPath.c_str());
+         dump_num++;
+         return;
+      }
+   }
+   LOG_DEBUG("Unable to find or open directory " + CPC.sdump_dir + " when trying to take a screenshot.");
```
Comment (ColinPitrat) [inline src/cap32.cpp]: Rather use LOG_ERROR ?

File: src/cap32.cpp [hunk 3]
```diff
@@ -3540,6 +3563,14 @@ int cap32_main (int argc, char **argv)
                            audio_resume();
                            break;
 
+                        case CAP32_SCRNDUMP:
+#ifdef HAVE_PNG
+                           CPC.paused |= 1;
```
Comment (ColinPitrat) [inline src/cap32.cpp]: Pausing is not necessary. Caprice32 is not multithreaded so nothing else will happen during dumpScreen.

File: src/gui/src/CapriceAbout.cpp [hunk 4]
```diff
@@ -17,7 +17,7 @@ CapriceAbout::CapriceAbout(const CRect& WindowRect, CWindow* pParent, CFontEngin
 	m_pMessageLabel = new CLabel(CPoint(5, 70), this, "Version 4.3.0");
 	m_pMessageLabel = new CLabel(CPoint(5, 90), this, "F1 - Menu / Pause");
 	m_pMessageLabel = new CLabel(CPoint(5, 100), this, "F2 - Fullscreen");
-	m_pMessageLabel = new CLabel(CPoint(5, 110), this, "F3 - ");
+	m_pMessageLabel = new CLabel(CPoint(5, 110), this, "F3 - Save screenshot");
 	//m_pMessageLabel = new CLabel(CPoint(5, 120), this, "F3 - Load/Save disk/tape/snapshot/screenshot");
```
Comment (ColinPitrat) [inline src/gui/src/CapriceAbout.cpp]: You can remove this comment (not sure why I kept it, the coordinates are not even right ...)

File: src/savepng.cpp [hunk 5]
```diff
@@ -0,0 +1,155 @@
+/*
+ * SDL_SavePNG -- libpng-based SDL_Surface writer.
```
Comment (ColinPitrat) [inline src/savepng.cpp]: You could add here and in header:
Code taken from driedfruit: https://github.com/driedfruit/SDL_SavePNG

File: src/cap32.cpp (hunk 1 above)
Comment (sebhz) [inline src/cap32.cpp]: Neat. Done.

File: src/cap32.cpp (hunk 2 above)
Comment (sebhz) [inline src/cap32.cpp]: Sure. Done.

File: src/cap32.cpp (hunk 3 above)
Comment (sebhz) [inline src/cap32.cpp]: OK. Done.

File: src/savepng.cpp (hunk 5 above)
Comment (sebhz) [inline src/savepng.cpp]: done.

**Comment by ColinPitrat** [comment]
Thanks, this is pretty nice !

**Pull request merged**